sbi/samplers/vi/vi_divergence_optimizers.py

```diff
@@ -0,0 +1,710 @@
+from __future__ import annotations
+
+import torch
+from torch.distributions import Distribution
+from torch import Tensor
+from torch import nn
+from torch.optim import Adam, Optimizer
+from torch.optim.lr_scheduler import ExponentialLR
+
+from typing import Optional, List, Tuple, Callable, Union
+from sbi.types import Array
+
+
+import numpy as np
+from abc import ABC, abstractmethod
+from copy import deepcopy
+
+from sbi.inference.potentials.base_potential import BasePotential
+
+
+from .vi_utils import (
+    filter_kwrags_for_func,
+    make_object_deepcopy_compatible,
+    move_all_tensor_to_device,
+)
+
+from .vi_sampling import clamp_weights, paretto_smoothed_weights
+
+_VI_method = {}
+
+
+class DivergenceOptimizer(ABC):
+    """This is a wrapper round a PyTorch optimizer and scheduler, which will be used to
+    learn the variational distribution. It further contains some methods to evaluate
+    convergence and to recover a valid state if something wents wrong.
+
+    This class contains one abstract method '_generate_loss_function' which must be
+    implemented und determines the loss function used within variational inference."""
+
+    def __init__(
+        self,
+        potential_fn: BasePotential,
+        q: Distribution,
+        n_particles: int = 256,
+        clip_value: float = 5.0,
+        optimizer: "Optimizer" = Adam,
+        scheduler: object = ExponentialLR,
+        eps: float = 1e-5,
+        **kwargs,
```

Comment:
Please expose the kwargs as individual arguments

Reply:
This is also kinda hard, because they change depending of the choice of scheduler/optimizer.

Reply:
ah, good point.

Reply:
I guess I will limit to choice to some schedulers. And make the documentation/typing more explicit.